ext/bcmath/libbcmath/src/str2num.c

```diff
@@ -57,77 +60,91 @@ bool bc_str2num(bc_num *num, char *str, size_t scale)
 	while (*ptr == '0') {
 		ptr++;
 	}
+	integer_ptr = ptr;
 	/* digits before the decimal point */
 	while (*ptr >= '0' && *ptr <= '9') {
 		ptr++;
 		digits++;
 	}
 	/* decimal point */
-	if (*ptr == '.') {
-		ptr++;
+	decimal_point = (*ptr == '.') ? ptr : NULL;
+
+	/* If a string other than numbers exists */
+	if (!decimal_point && *ptr != '\0') {
+		goto fail;
 	}
-	/* digits after the decimal point */
-	while (*ptr >= '0' && *ptr <= '9') {
-		if (*ptr == '0') {
-			trailing_zeros++;
-		} else {
-			trailing_zeros = 0;
+
+	/* search and validate fractional end if exists */
+	if (decimal_point) {
+		/* search */
+		fractional_ptr = fractional_end = decimal_point + 1;
+		if (*fractional_ptr == '\0') {
+			goto after_fractional;
 		}
-		ptr++;
-		strscale++;
-	}
 
-	if (trailing_zeros > 0) {
-		/* Trailing zeros should not take part in the computation of the overall scale, as it is pointless. */
-		strscale = strscale - trailing_zeros;
+		/* validate */
+		while (*fractional_ptr >= '0' && *fractional_ptr <= '9') {
+			fractional_end = *fractional_ptr != '0' ? fractional_ptr + 1 : fractional_end;
+			fractional_ptr++;
+		}
+		if (*fractional_ptr != '\0') {
+			/* invalid num */
+			goto fail;
+		}
+		fractional_ptr = decimal_point + 1;
+
+		str_scale = fractional_end - fractional_ptr;
```

Comment:
This may need a comment.

Reply:
Added https://github.com/php/php-src/pull/14076/commits/386cb7f0ac1a90542ba02a4a3dba56ad6171775d